### PR TITLE
Colab for a simple DPO run 

### PR DESCRIPTION
**Pull request by ritwikraha**
This notebook:

- Creates a simple notebook for a DPO run
- Implements the dpo loss
- Fetches the GPT-2 model and fine tunes it on a data (json)
- Saves the model and tokenizer

What is needed?

- A proper data pipeline
- Proper inference pipeline

### CODE REVIEW

**Comment by ariG23498** [comment]
Some pointers:
1. Dataset: Right now the dummy dataset (Preference Dataset) is a good first step. Later when we use a dataset from the Hub, we have to create the `collate_function` and pass that to the `DataLoader`. This means, we do not have to pay too much heed to the `Dataset` at the moment.
2. Tokenization inside collate: I think once we have a dataset to work on, we should batch tokenize the samples rather than doing it one by one. This would increase throughput.
3. Add special tokens: GPT2 does not have a varied amount of speical tokens, so we can add some and resize the embeddings. Here is how I would have done it
```py
model_id = "openai-community/gpt2"
tokenizer = AutoTokenizer.from_pretrained(model_id)
model = AutoModelForCausalLM.from_pretrained(model_id)

# Add special tokens to the tokenizer (notice eos_token is already present)
special_tokens_dict = {
    "bos_token": "<|beginoftext|>",
    "pad_token": "<|pad|>",
}
num_added_toks = tokenizer.add_special_tokens(special_tokens_dict)
model.resize_token_embeddings(len(tokenizer)) # Importat step to resize the embeddings of the model
```
4. Format the inputs: We can format the inputs with `bos_token` and `eos_token`, or better yet, have a chat template to go with the model, but that comes much later.
5. Each transformers model has a `label` parameter, which when provided, the model computes the loss between the prediction and the grouth truth labels. This would do away with logit shifting and masking of pad tokens.

File: nanorl/algorithms/dpo/data.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+class DPODataset(Dataset):
+    def __init__(self, data: List[Dict]):
+        self.data = data
+
+    def __len__(self):
+        return len(self.data)
+
+    def __getitem__(self, idx):
+        return self.data[idx]
```
Comment (ariG23498) [inline nanorl/algorithms/dpo/data.py:14]: This would be replaces with a HF dataset

`ds = load_dataset("hf_data")`

File: nanorl/algorithms/dpo/data.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        rejected_texts,
+        max_length=max_length,
+        padding=True,
+        truncation=True,
+        return_tensors="pt",
+    )
+
+    prompt_tokens = tokenizer(
+        prompts, padding=False, truncation=True, add_special_tokens=True
+    )
```
Comment (ariG23498) [inline nanorl/algorithms/dpo/data.py:42]: Will seem redundant, but is required to compute the length of tokens in the prompts. This is used later to determine where to ignore as the chose and rejected both consist of the prompt which is not needed in the loss computation.

File: nanorl/data/dummy_preference.py
Comment (ariG23498) [inline nanorl/data/dummy_preference.py:1]: @ritwikraha this file needs to be changed with the process you take to build the synthetic dataset.

Comment (ritwikraha) [reply nanorl/data/dummy_preference.py:1]: got it

File: scripts/train_dpo.py
Comment (ritwikraha) [inline scripts/train_dpo.py:1]: Consider moving this to the root directory to access the dpo module easily

File: nanorl/algorithms/dpo/train.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+            loss = loss / config.gradient_accumulation_steps
+            loss.backward()
+
+            epoch_loss += loss.item() * config.gradient_accumulation_steps
+            epoch_accuracy += accuracy.item()
+            num_batches += 1
+
+            if (batch_idx + 1) % config.gradient_accumulation_steps == 0:
+                torch.nn.utils.clip_grad_norm_(policy_model.parameters(), 1.0)
```
Comment (ritwikraha) [inline nanorl/algorithms/dpo/train.py:112]: nitpick: consider having this in the imports `torch.nn.utils.clip_grad_norm_`

File: nanorl/algorithms/dpo/model.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+        shift_logits = logits[:, :-1, :].contiguous()
+        shift_labels = labels[:, 1:].contiguous()
+
+        log_probs = F.log_softmax(shift_logits, dim=-1)
+        gathered_log_probs = torch.gather(
+            log_probs, dim=-1, index=shift_labels.unsqueeze(-1)
+        ).squeeze(-1)
+
+        mask = (shift_labels != -100) & (shift_labels != model.config.pad_token_id)
```
Comment (ritwikraha) [inline nanorl/algorithms/dpo/model.py:18]: good
consider setting this 100 in the config

`token_cut_off = 100`

**Review by ritwikraha — COMMENTED** [review]
LGTM, please go through the comments
Some torch imports can be better handled

Do you want `max_length = max_length` or max_length = `max_lenth*3` while the first simply chooses the max number of tokens, the latter is a way to ensure that the chosen tokens resemble words (each word is approx 3 token). Doing this ensures a more polite cut off

File: nanorl/algorithms/dpo/model.py (hunk 4 above)
Comment (ariG23498) [reply nanorl/algorithms/dpo/model.py:18]: It is pretty well defined in the eco-system of torch to use `-100` to ignore.

image

File: nanorl/algorithms/dpo/train.py (hunk 3 above)
Comment (ritwikraha) [reply nanorl/algorithms/dpo/train.py:112]: lets keep as is

File: nanorl/algorithms/dpo/model.py (hunk 4 above)
Comment (ritwikraha) [reply nanorl/algorithms/dpo/model.py:18]: cool, lets keep as is
